hybrid/ddc_bounds.c

```diff
@@ -0,0 +1,21 @@
+#include "../include/common.h"
+#include "include/utils.h"
+#include <stdint.h>
+#include <stdlib.h>
+#if !defined(__aarch64__)
+#error "This example is Morello only"
+#endif
+
+int main(){
+    int32_t array[16] = {0};
+	int32_t *array_ptr = &array;
+    pp_cap(cheri_ddc_get());
+    int32_t *__capability array_cap = (int32_t *__capability) array_ptr;
```

Comment:
That casts an `int32_t*` to an `int32_t* __capability`, which I think just derives the metadata from the DDC. That means that this first bit just sets the address of the DDC. Am I mistaken? I don't think that derivation sets the bounds.

Reply:
So, casting here is not sufficient...

Reply:
I don't think you want `cheri_cap_build` here.

----

Loosely: `cheri_cap_build(x, y)` takes one bit pattern, `y`, that looks like a capability (but perhaps isn't tagged), and a valid capability `x`, and derives a new capability from `x` that looks like `y`, but valid and unsealed. It doesn't give you any capability you couldn't have already derived (since you have to have `x` in the first place).

This allows you to do two things, broadly:
* If you have a capablity `x`, and a sealed capability `y` that is otherwise a subset of `x`, you can unseal `y` (using `x` as credentials instead of an `UNSEAL` permission).
* If you have a capability `x` and an _invalid_ capability `y` that would have been a subset of `x` if it were valid, you can derive a capability from `x` that looks like `y` with its tag set.

Both operations could be achieved by a series of individual operations on `x` (e.g. `cheri_bounds_set()` and so on).

Reply:
> It depends on what you want to do. What bounds do you expect `array_cap` to have? `array_ptr` is a just a plain address, and the C type has no size information. I doubt that the compiler will attempt to preserve the bounds of `array` across those casts (though in purecap it would because `&array` would itself be a capability).
> 
> If your intention is to restrict the bounds of `array_cap` to the array itself, then you probably need to use `cheri_bounds_set` to do that explicitly. (Check that, though, because I haven't.)

If I try to set the bounds on the capability, when I do a `write_ddc` I get a `Capability bounds fault`. So, there is something I haven't got right here...